model/task_history.go

```diff
@@ -776,68 +776,48 @@ func testHistoryV2Results(params *TestHistoryParameters) ([]task.Task, error) {
 
-	// to join the test results, merge test results for all the tasks...
-	if len(params.TestNames) == 0 {
-		out := []task.Task{}
-		taskChan := make(chan task.Task, len(tasks))
-		for _, t := range tasks {
-			taskChan <- t
-		}
-		close(taskChan)
-		wg := sync.WaitGroup{}
-		wg.Add(numQueryThreads)
-		resultChan := make(chan task.Task)
-		for i := 0; i < numQueryThreads; i++ {
-			go func() {
-				defer wg.Done()
-				for t := range taskChan {
-					err := t.MergeNewTestResults() //nolint
-					if err != nil {
-						grip.Error(errors.Wrapf(err, "error merging test results for task %s", t.Id))
-						continue
-					}
-					for j := len(t.LocalTestResults) - 1; j >= 0; j-- {
-						result := t.LocalTestResults[j]
-						if len(params.TestStatuses) > 0 && !util.StringSliceContains(params.TestStatuses, result.Status) {
-							t.LocalTestResults = append(t.LocalTestResults[:j], t.LocalTestResults[j+1:]...)
-						}
+	// to join the test results, merge test results for all the tasks
+	out := []task.Task{}
+	taskChan := make(chan task.Task, len(tasks))
+	for _, t := range tasks {
+		taskChan <- t
+	}
+	close(taskChan)
+	wg := sync.WaitGroup{}
+	wg.Add(numQueryThreads)
+	resultChan := make(chan task.Task)
+	for i := 0; i < numQueryThreads; i++ {
+		go func() {
+			defer wg.Done()
+			for t := range taskChan {
+				err := t.MergeNewTestResults() //nolint
+				if err != nil {
+					grip.Error(errors.Wrapf(err, "error merging test results for task %s", t.Id))
+					continue
+				}
+				// strip out any test results that don't match input params
+				for j := len(t.LocalTestResults) - 1; j >= 0; j-- {
+					result := t.LocalTestResults[j]
+					if (len(params.TestStatuses) > 0 && !util.StringSliceContains(params.TestStatuses, result.Status)) ||
+						(len(params.TestNames) > 0 && !util.StringSliceContains(params.TestNames, result.TestFile)) {
+						t.LocalTestResults = append(t.LocalTestResults[:j], t.LocalTestResults[j+1:]...)
 					}
-					resultChan <- t
 				}
-			}()
-		}
-		doneChan := make(chan bool)
-		go func() {
-			defer close(doneChan)
-			for t := range resultChan {
-				out = append(out, t)
+				resultChan <- t
 			}
 		}()
-		wg.Wait()
-		close(resultChan)
-		<-doneChan
-		return out, nil
-	}
-	//... or if test names are specified, do a query on testresults and match up executions
-	testsQuery := formQueryFromTests(params)
-	results, err := testresult.Find(db.Query(testsQuery))
-	if err != nil {
-		return nil, err
 	}
-	for i := 0; i < len(results); i++ {
-		result := results[i]
-	taskLoop:
-		for j := 0; j < len(tasks); j++ {
-			t := tasks[j]
-			if result.TaskID == t.Id && result.Execution == t.Execution {
-				t.LocalTestResults = append(t.LocalTestResults, task.ConvertToOld(&result))
-				tasks[j] = t
-				break taskLoop
-			}
+	doneChan := make(chan bool)
+	go func() {
+		defer close(doneChan)
+		for t := range resultChan {
+			out = append(out, t)
 		}
-	}
-
-	return tasks, nil
+	}()
+	wg.Wait()
+	close(resultChan)
+	<-doneChan
```

Comment:
I think if you buffer the output channel, you can get rid of the second command?

Reply:
I don't know the size of the output channel ahead of time